### ISSUE TITLE
benchmarks: surpring ocaml/ocaml benchmark results

### ISSUE THREAD

**Issue by gasche** [ocaml/infrastructure#80]
I had a look at the benchmark results for ocaml/ocaml today, and something is surprising: it appears that a certain change on the OCaml compiler made menhir type-checking time go from 4s to 287s. I think that this is a miscomputation of the benchmark scripts, but I am not sure, so reporting here.

## Step 1: find the relevant change

On https://autumn.ocamllabs.io/ocaml/ocaml/branch/trunk?worker=autumn&image=ocaml%2Fopam%3Adebian-11-ocaml-4.14 , if use the calendar widget in the top right to select the date range "2023-07-01 - 2023-07-17", I see a graph going up very strongly in the "projects > menhir" category -- `typing` goes from 4s to 287s.

**Side questions:**
1. can I get a permalink with these date ranges, so that I can share the URL with the dates selected instead of having other people enter the same range again?
2. why is the output of all graphs grayed out when I select this graph range?

## Step 1.99: reproduce the result

The ocaml/ocaml PR that was merged around this time was https://github.com/ocaml/ocaml/pull/12236 . It sounds *plausible* that this PR would have an unplanned impact on compile times, including type-checking.

However, I have failed to reproduce any sizeable difference in menhir build times.

## Step 2: study the logs

Immediately on the right of the "menhir" name in the "projects" category there is a small question icon, which points to the following URL: http://autumn.ocamllabs.io:8081/job/2023-07-17/093001-ocluster-build-338a97#L75689-L75810
From the URL I assume that this is the build log on 2023-07-17, so after the change that (supposedly) blew up type-checking time for menhir.

**Side question:** this gets me the log of the end commit of the interval I selected. How do I get the log of the beginning commit, or another commit in between?

Comparing the build logs from 2023-06-30 (before the change) and from 2023-07-17 (after the change), I see the following:

On 06-30:
```
[NOTE] menhir is not installed.

Nothing to do.
The following actions will be performed:
  - install menhirLib 20220210 [required by menhir]
  - install menhirSdk 20220210 [required by menhir]
  - install menhir    20220210
===== 3 to install =====
```

On 07-17:

```
[NOTE] menhir is not installed.

Nothing to do.
The following actions will be performed:
  - install   ocaml-options-vanilla 1
  - remove    ocaml-variants        5.2.0+trunk*   [conflicts with ocaml]
  - install   ocaml-base-compiler   5.0.0          [required by ocaml]
  - recompile ocaml-config          3              [uses ocaml-variants]
  - downgrade ocaml                 5.2.0 to 5.0.0 [required by menhir]
  - recompile dune                  3.4.1          [uses ocaml]
  - recompile base-domains          base           [uses ocaml]
  - install   menhirSdk             20220210       [required by menhir]
  - install   menhirLib             20220210       [required by menhir]
  - recompile stdlib-shims          0.3.0          [uses ocaml]
  - install   base-nnp              base
  - install   menhir                20220210
  - recompile ocamlgraph            2.0.0          [uses ocaml]
===== 6 to install | 5 to recompile | 1 to downgrade | 1 to remove =====
```

My best guess as to what is going on is that the benchmarking script somehow counts all that extra work (downgrading OCaml, recompiling ocamlgraph etc.) as part of the `menhir` times, and somehow gets an over-inflated figure for type-checking time.

Note that the high build time for `menhir` remain in the latest build logs for ocaml/ocaml ( http://autumn.ocamllabs.io:8081/job/2023-11-13/093001-ocluster-build-af9d09#L77399-L77520 ), it seems that the issue that makes it downgrade to an older OCaml version is still there.

(cc @Octachron @punchagan)

**Comment by punchagan** [comment]
> My best guess as to what is going on is that the benchmarking script somehow counts all that extra work (downgrading OCaml, recompiling ocamlgraph etc.) as part of the menhir times, and somehow gets an over-inflated figure for type-checking time.

I took a quick look at the build logs and it looks like your hypothesis may be right.  

The benchmarking scripts get the timing information from the debug output from opam install with `OCAMLPARAM="_,timings=1"`

```sh
OCAMLPARAM="_,timings=1" opam install -b --verbose -y "${project}=${version}" 2>&1 | tee -a build.log
```
If an older version of `ocaml` is being reinstalled, it seems like the timing output would be polluted with the timing information from the ocaml and other dependency installation. 

The benchmarking scripts fix a version of the projects being used to get the benchmarks timing information, and it looks like the version of `menhir` frozen in those scripts is not compatible with the latest `trunk`. I'll look further into this and get back with a fix.

**Comment by punchagan** [comment]
@gasche Thank you for taking the effort to get to the bottom of this and the detailed issue! Appreciate your detailed issue and initial work to help with debugging the problem.

> My best guess as to what is going on is that the benchmarking script somehow counts all that extra work (downgrading OCaml, recompiling ocamlgraph etc.) as part of the menhir times, and somehow gets an over-inflated figure for type-checking time.

I took a quick look at the build logs and it looks like your hypothesis may be right.  

The benchmarking scripts get the timing information from the debug output from opam install with `OCAMLPARAM="_,timings=1"`

```sh
OCAMLPARAM="_,timings=1" opam install -b --verbose -y "${project}=${version}" 2>&1 | tee -a build.log
```
If an older version of `ocaml` is being reinstalled, it seems like the timing output would be polluted with the timing information from the ocaml and other dependency installation. 

The benchmarking scripts fix a version of the projects being used to get the benchmarks timing information, and it looks like the version of `menhir` frozen in those scripts is not compatible with the latest `trunk`. I'll look further into this and get back with a fix.

**Comment by punchagan** [comment]
Hello @gasche,

Firstly, thanks again for your thorough investigation and for bringing this issue to our attention.

Regarding the timing information, as I mentioned before we calculate it using the output from the following command:

```bash
OCAMLPARAM="_,timings=1" opam install -b --verbose -y "${project}=${version}"
```

For the purpose of maintaining consistent benchmarks across different versions of OCaml for each project, we [pin](https://github.com/punchagan/ocaml-benching/blob/clean-bench.sh/bench.sh#L178-L181) the versions. In the case of `menhir`, it is pinned at `20220210`. However, this version of `menhir` seems incompatible with the latest OCaml trunk, triggering a downgrade to an older version of OCaml.

To address the current issue and correct the benchmarks, we could pin `menhir` to a more recent version.

That said, I'm pondering over a suitable long-term approach for these benchmarks. Changing the project version (in other words the code being compiled) might undermine the integrity of the benchmark. I would welcome any suggestions or ideas you might have on how best to navigate this.

>     1. can I get a permalink with these date ranges, so that I can share the URL with the dates selected instead of having other people enter the same range again?

Unfortunately, that feature is not available at the moment. I've created an [issue](https://github.com/ocurrent/current-bench/issues/488) for this

>     2. why is the output of all graphs grayed out when I select this graph range?

It appears to be a bug. The graying out is intended to indicate when the latest build fails, allowing users to still view the graphs while signaling an issue with the latest commit. However, when a specific time window is selected that does not include the latest commit, the UI incorrectly grays out the graph. I've created an [issue](https://github.com/ocurrent/current-bench/issues/489) to track this.

> **Side question:** this gets me the log of the end commit of the interval I selected. How do I get the log of the beginning commit, or another commit in between?

The direct links to access build logs for specific commits are currently not available, which is indeed a limitation. Here's an [issue](https://github.com/ocurrent/current-bench/issues/490)

Thank you again for your valuable input and assistance in identifying these issues.

**Comment by gasche** [comment]
Thanks for the detailed feedback.

I see two ways to approach this -- trying to upgrade the benchmark from an old version of Menhir, say old-menhir, to a newer version, say new-menhir:

- You can try to hide the change by recompiling old compiler commits with new-menhir, so that all results use the same implementation. Downside: a lot of extra compute to do, and also in the long term this may be limited if recompiling with older OCaml versions becomes harder to reproduce.
- Or you can show them as separate benchmarks (with data collection for old-Menhir stopping at one point), so that people are forced to notice the difference and interpret the results carefully.

**Comment by punchagan** [comment]
> Or you can show them as separate benchmarks (with data collection for old-Menhir stopping at one point), so that people are forced to notice the difference and interpret the results carefully.

I've implemented a change to show them as separate benchmarks in tarides/ocaml-benching#3 . I've also fixed the incorrect downgrade to older OCaml versions issue. 

The issues of the benchmark being surprising (read as: incorrect/meaningless) should be fixed now. 

I think this issue can be closed, since the issues with the `current-bench` UI are being tracked in the project's repo. 

Thanks again, @gasche, for you detailed bug report and bringing our attention to this issue!
